src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs

```diff
@@ -242,7 +244,7 @@ private void RestoreQueues()
 
         private void OnEndOfShift(object sender, EventArgs ea)
         {
-            if (_isolationLevel > 0)
+            if (_isolationLevel > 0 && Shifts.All(q => !q.HasWork))
```

Comment:
I don't actually understand what happened here. I know I added code to fix an earlier bug that made exactly this test using a different, non-Linq implementation. But that change seems to have disappeared!

Anyway, this is definitely needed.

Reply:
However... I think there is more to it. It's possible for all the queues to be empty while there is still unfinished work because some fixture is running but has not yet queued its child tests. This is handled in the immediately following code for the purpose of not terminating the appication...

```C#
            // Shift has ended but all work may not yet be done
            while (_topLevelWorkItem.State != WorkItemState.Complete)
            {
                // This will fail if there is no work - all queues empty.
                // In that case, we just continue the loop until either
                // a shift is started or all the work is complete.
                if (StartNextShift())
                    return;
            }
```

I don't have time to step through it right now, but I'm worried that there's a race condition buried here, where a new test gets queued just as you are making your test.

Reply:
You are using a different condition here than in your guard. I tried switching this to `Queues.All(q => q.IsEmpty)` and it works some of the time, but we need to add a null check when dequeuing and the guard failed one time making me think there is a race condition.

I also tried by changing the guard to `Shifts.All(q => !q.HasWork)` but that doesn't work.

I am running out of time today, I will try to get back to looking at it tonight.

Reply:
@CharliePoole - yes, that sounds feasible. I'm not totally sure how to deal with that one. The whole saving and restoring queues logic seems difficult to multi-thread - it wasn't entirely clear what the side-effects of dealing with https://github.com/nunit/nunit/issues/2428 would be here either.

@rprouse - Aha! I umm'd and arr'd over which condition to use, and seem to have ended up with one of each! I thought the two should be functionally equivalent here - interested as to what's not working by switching them around...

Reply:
> but we need to add a null check when dequeuing

@rprouse - what did you mean by that? Are you possibly seeing https://github.com/nunit/nunit/issues/2428? That needs to be dealt with as a separate issue, imo.